### ISSUE TITLE
Test failures when upgrading Clojure from 1.8 to 1.9 (spec conformance)

### ISSUE THREAD

**Issue by CyrilBrulebois** [puppetlabs/comidi#36]
Hi,

I'm trying to get PuppetDB in a suitable state for the next Debian stable release (Buster, Debian 10), which will be frozen in a few weeks. The `puppetdb` package depends on `libcomidi-clojure`, which can no longer be built from source, due to test issues:

```
dh_auto_test
(cd test && find . -name "*.clj" | \
	xargs --verbose clojure -cp /comidi-clojure-0.3.1/comidi.jar:/usr/share/java/clojure.jar:/usr/share/java/clj-time.jar:/usr/share/java/bidi.jar:/usr/share/java/compojure.jar:/usr/share/java/prismatic-schema.jar:/usr/share/java/kitchensink.jar)
clojure -cp /comidi-clojure-0.3.1/comidi.jar:/usr/share/java/clojure.jar:/usr/share/java/clj-time.jar:/usr/share/java/bidi.jar:/usr/share/java/compojure.jar:/usr/share/java/prismatic-schema.jar:/usr/share/java/kitchensink.jar ./puppetlabs/comidi_test.clj 
Exception in thread "main" clojure.lang.ExceptionInfo: Call to clojure.core/ns did not conform to spec:
In: [1] val: ((require [clojure.test :refer :all] [puppetlabs.comidi :as comidi :refer :all] [schema.test :as schema-test] [schema.core :as schema] [clojure.zip :as zip] [bidi.bidi :as bidi])) fails at: [:args] predicate: (cat :docstring (? string?) :attr-map (? map?) :clauses :clojure.core.specs/ns-clauses),  Extra input
:clojure.spec/args  (puppetlabs.comidi-test (require [clojure.test :refer :all] [puppetlabs.comidi :as comidi :refer :all] [schema.test :as schema-test] [schema.core :as schema] [clojure.zip :as zip] [bidi.bidi :as bidi]))
 #:clojure.spec{:problems [{:path [:args], :reason "Extra input", :pred (cat :docstring (? string?) :attr-map (? map?) :clauses :clojure.core.specs/ns-clauses), :val ((require [clojure.test :refer :all] [puppetlabs.comidi :as comidi :refer :all] [schema.test :as schema-test] [schema.core :as schema] [clojure.zip :as zip] [bidi.bidi :as bidi])), :via [], :in [1]}], :args (puppetlabs.comidi-test (require [clojure.test :refer :all] [puppetlabs.comidi :as comidi :refer :all] [schema.test :as schema-test] [schema.core :as schema] [clojure.zip :as zip] [bidi.bidi :as bidi]))}, compiling:(/comidi-clojure-0.3.1/test/./puppetlabs/comidi_test.clj:1:1)
	at clojure.lang.Compiler.load(Compiler.java:7442)
	at clojure.lang.Compiler.loadFile(Compiler.java:7368)
	at clojure.main$load_script.invokeStatic(main.clj:277)
	at clojure.main$script_opt.invokeStatic(main.clj:337)
	at clojure.main$script_opt.invoke(main.clj:332)
	at clojure.main$main.invokeStatic(main.clj:423)
	at clojure.main$main.doInvoke(main.clj:386)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.lang.Var.applyTo(Var.java:700)
	at clojure.main.main(main.java:37)
Caused by: clojure.lang.ExceptionInfo: Call to clojure.core/ns did not conform to spec:
In: [1] val: ((require [clojure.test :refer :all] [puppetlabs.comidi :as comidi :refer :all] [schema.test :as schema-test] [schema.core :as schema] [clojure.zip :as zip] [bidi.bidi :as bidi])) fails at: [:args] predicate: (cat :docstring (? string?) :attr-map (? map?) :clauses :clojure.core.specs/ns-clauses),  Extra input
:clojure.spec/args  (puppetlabs.comidi-test (require [clojure.test :refer :all] [puppetlabs.comidi :as comidi :refer :all] [schema.test :as schema-test] [schema.core :as schema] [clojure.zip :as zip] [bidi.bidi :as bidi]))
 {:clojure.spec/problems [{:path [:args], :reason "Extra input", :pred (cat :docstring (? string?) :attr-map (? map?) :clauses :clojure.core.specs/ns-clauses), :val ((require [clojure.test :refer :all] [puppetlabs.comidi :as comidi :refer :all] [schema.test :as schema-test] [schema.core :as schema] [clojure.zip :as zip] [bidi.bidi :as bidi])), :via [], :in [1]}], :clojure.spec/args (puppetlabs.comidi-test (require [clojure.test :refer :all] [puppetlabs.comidi :as comidi :refer :all] [schema.test :as schema-test] [schema.core :as schema] [clojure.zip :as zip] [bidi.bidi :as bidi]))}
	at clojure.core$ex_info.invokeStatic(core.clj:4725)
	at clojure.spec$macroexpand_check.invokeStatic(spec.clj:697)
	at clojure.spec$macroexpand_check.invoke(spec.clj:686)
	at clojure.lang.AFn.applyToHelper(AFn.java:156)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.lang.Var.applyTo(Var.java:700)
	at clojure.lang.Compiler.macroexpand1(Compiler.java:6816)
	at clojure.lang.Compiler.macroexpand(Compiler.java:6888)
	at clojure.lang.Compiler.eval(Compiler.java:6962)
	at clojure.lang.Compiler.load(Compiler.java:7430)
	... 9 more
debian/rules:32: recipe for target 'override_dh_auto_test' failed
```

It was reported in [Debian#889125](https://bugs.debian.org/889125), and I've spent some time investigating the causes, leading me to narrow it down to the Clojure upgrade from 1.8.0-3 to 1.9.0~alpha15-1. I've documented all my findings in [message #12](https://bugs.debian.org/889125#12) of that bug report, but you'll find below the Clojure specific parts (comidi, bidi).

I've tried upgrading comidi to the latest upstream release (0.3.2), which didn't change anything; I haven't learned about leiningen enough just yet but I've spotted the bidi upgrade, and tried to upgrade the debian package from 2.1.2 to 2.1.3 (given 4e5c91b8d57b69cbfed02c728922c1f8ce071f9d, even if that seemed to be irrelevant at first), which led to no better results.

Maybe fabc6d83331fbf5af56e6f76f917ef4586dbf294 and/or 9a5bfdf4f16a0b7455556591c898fb2071e4c156 might help, if I understood what they are achieving exactly and how to translate that at the debian package level? But at least from the commit messages, they don't look like obvious fixes for the spec conformance issue we are seeing when trying to run tests?

Thanks for any help/guidance.

Season's greetings,
Cyril.

**Comment by puredanger** [comment]
In case it helps, the actual bug here is that the ns form in puppetlabs.comidi-test uses `require` instead of `:require`. The former accidentally worked in Clojure 1.8 but specs were added to mark this as invalid syntax in Clojure 1.9.

**Comment by CyrilBrulebois** [comment]
Thanks so much, @puredanger! I've just tested this fix successfully and opened MR #37 accordingly.
